File: gsonpath-compiler/src/main/java/gsonpath/generator/adapter/AutoGsonAdapterPropertiesFactory.java
package gsonpath.generator.adapter;

import com.google.gson.FieldNamingPolicy;
import gsonpath.*;
import gsonpath.model.GsonFieldNamingPolicyFactory;

import java.util.HashSet;
import java.util.Set;

class AutoGsonAdapterPropertiesFactory {

    AutoGsonAdapterProperties create(AutoGsonAdapter autoGsonAnnotation, GsonPathDefaultConfiguration defaultsAnnotation,
                                     boolean isInterface) throws ProcessingException {

        boolean fieldsRequireAnnotation = autoGsonAnnotation.ignoreNonAnnotatedFields().booleanValue;
        boolean serializeNulls = autoGsonAnnotation.serializeNulls().booleanValue;
        char flattenDelimiter = autoGsonAnnotation.flattenDelimiter().value();
        GsonFieldValidationType gsonFieldValidationType = autoGsonAnnotation.fieldValidationType();
        PathSubstitution[] pathSubstitutions = autoGsonAnnotation.substitutions();

        // Validate the path substitutions. Duplicate keys are not allowed.
        Set<String> pathSubstitutionKeys = new HashSet<>(pathSubstitutions.length);
        for (PathSubstitution pathSubstitution : pathSubstitutions) {
            if (pathSubstitutionKeys.contains(pathSubstitution.original())) {
                throw new ProcessingException("PathSubstitution original values must be unique");
            }
            pathSubstitutionKeys.add(pathSubstitution.original());
        }

        // We want to translate the Gson Path 'GsonPathFieldNamingPolicy' enum into the standard Gson version.
        FieldNamingPolicy gsonFieldNamingPolicy = new GsonFieldNamingPolicyFactory().getPolicy(autoGsonAnnotation.fieldNamingPolicy());

        if (defaultsAnnotation != null) {

            // Inherit 'ignoreNonAnnotatedFields'
            if (autoGsonAnnotation.ignoreNonAnnotatedFields().inheritDefaultIfAvailable) {
                fieldsRequireAnnotation = defaultsAnnotation.ignoreNonAnnotatedFields();
            }

            // Inherit 'serializeNulls'
            if (autoGsonAnnotation.serializeNulls().inheritDefaultIfAvailable) {
                serializeNulls = defaultsAnnotation.serializeNulls();
            }

            // Inherit 'flattenDelimiter'
            if (autoGsonAnnotation.flattenDelimiter().inheritDefaultIfAvailable()) {
                flattenDelimiter = defaultsAnnotation.flattenDelimiter();
            }

            // Inherit 'fieldNamingPolicy'
            if (autoGsonAnnotation.fieldNamingPolicy().equals(GsonPathFieldNamingPolicy.IDENTITY_OR_INHERIT_DEFAULT_IF_AVAILABLE)) {
                gsonFieldNamingPolicy = defaultsAnnotation.fieldNamingPolicy();
            }

            // Inherit 'fieldValidationType'
            if (gsonFieldValidationType.equals(GsonFieldValidationType.NO_VALIDATION_OR_INHERIT_DEFAULT_IF_AVAILABLE)) {
                gsonFieldValidationType = defaultsAnnotation.fieldValidationType();
            }
        }

        // Interfaces must use field validation to prevent issues with primitives.
        if (isInterface) {
            switch (gsonFieldValidationType) {
                case NO_VALIDATION:
                case NO_VALIDATION_OR_INHERIT_DEFAULT_IF_AVAILABLE:
                    gsonFieldValidationType = GsonFieldValidationType.VALIDATE_EXPLICIT_NON_NULL;
                    break;
            }
        }

        return new AutoGsonAdapterProperties(fieldsRequireAnnotation, flattenDelimiter, serializeNulls,
                autoGsonAnnotation.rootField(), gsonFieldValidationType, gsonFieldNamingPolicy, pathSubstitutions);
    }

}

File: gsonpath-compiler/src/main/java/gsonpath/generator/standard/AutoGsonAdapterPropertiesFactory.kt
package gsonpath.generator.standard

import com.google.gson.FieldNamingPolicy
import gsonpath.*
import gsonpath.ProcessingException
import gsonpath.model.GsonFieldNamingPolicyFactory

internal class AutoGsonAdapterPropertiesFactory {

    @Throws(ProcessingException::class)
    fun create(autoGsonAnnotation: AutoGsonAdapter, defaultsAnnotation: GsonPathDefaultConfiguration?,
               isInterface: Boolean): AutoGsonAdapterProperties {

        val fieldsRequireAnnotation =
                if (defaultsAnnotation != null && autoGsonAnnotation.ignoreNonAnnotatedFields.inheritDefaultIfAvailable) {
                    // Inherit 'ignoreNonAnnotatedFields'
                    defaultsAnnotation.ignoreNonAnnotatedFields
                } else {
                    autoGsonAnnotation.ignoreNonAnnotatedFields.booleanValue
                }

        val serializeNulls =
                if (defaultsAnnotation != null && autoGsonAnnotation.serializeNulls.inheritDefaultIfAvailable) {
                    // Inherit 'serializeNulls'
                    defaultsAnnotation.serializeNulls
                } else {
                    autoGsonAnnotation.serializeNulls.booleanValue
                }

        val flattenDelimiter =
                if (defaultsAnnotation != null && autoGsonAnnotation.flattenDelimiter.inheritDefaultIfAvailable) {
                    // Inherit 'flattenDelimiter'
                    defaultsAnnotation.flattenDelimiter
                } else {
                    autoGsonAnnotation.flattenDelimiter.value
                }

        val gsonFieldValidationType =
                if (defaultsAnnotation != null && autoGsonAnnotation.fieldValidationType ==
                        GsonFieldValidationType.NO_VALIDATION_OR_INHERIT_DEFAULT_IF_AVAILABLE) {

                    if (isInterface) {
                        // Interfaces must use field validation to prevent issues with primitives.
                        when (defaultsAnnotation.fieldValidationType) {
                            GsonFieldValidationType.NO_VALIDATION,
                            GsonFieldValidationType.NO_VALIDATION_OR_INHERIT_DEFAULT_IF_AVAILABLE ->
                                GsonFieldValidationType.VALIDATE_EXPLICIT_NON_NULL

                            else ->
                                defaultsAnnotation.fieldValidationType
                        }
                    } else {
                        // Inherit 'fieldValidationType'
                        defaultsAnnotation.fieldValidationType
                    }

                } else {
                    if (isInterface) {
                        // Interfaces must use field validation to prevent issues with primitives.
                        when (autoGsonAnnotation.fieldValidationType) {
                            GsonFieldValidationType.NO_VALIDATION,
                            GsonFieldValidationType.NO_VALIDATION_OR_INHERIT_DEFAULT_IF_AVAILABLE ->
                                GsonFieldValidationType.VALIDATE_EXPLICIT_NON_NULL

                            else ->
                                autoGsonAnnotation.fieldValidationType
                        }
                    } else {
                        autoGsonAnnotation.fieldValidationType
                    }
                }

        val pathSubstitutions = autoGsonAnnotation.substitutions

        // Validate the path substitutions. Duplicate keys are not allowed.
        pathSubstitutions.fold(emptySet()) { set: Set<String>, pathSubstitution ->
            if (set.contains(pathSubstitution.original)) {
                throw ProcessingException("PathSubstitution original values must be unique")
            }
            set.plus(pathSubstitution.original)
        }

        // We want to translate the Gson Path 'GsonPathFieldNamingPolicy' enum into the standard Gson version.
        val gsonFieldNamingPolicy: FieldNamingPolicy =
                if (defaultsAnnotation != null && autoGsonAnnotation.fieldNamingPolicy ==
                        GsonPathFieldNamingPolicy.IDENTITY_OR_INHERIT_DEFAULT_IF_AVAILABLE) {

                    // Inherit 'fieldNamingPolicy'
                    defaultsAnnotation.fieldNamingPolicy
                } else {
                    GsonFieldNamingPolicyFactory().getPolicy(autoGsonAnnotation.fieldNamingPolicy)
                }

        return AutoGsonAdapterProperties(fieldsRequireAnnotation, flattenDelimiter, serializeNulls,
                autoGsonAnnotation.rootField, gsonFieldValidationType, gsonFieldNamingPolicy, pathSubstitutions)
    }

}
